src/test/java/com/squareup/javapoet/TestFiler.java

```diff
@@ -30,20 +31,28 @@
 import javax.tools.FileObject;
 import javax.tools.JavaFileManager;
 import javax.tools.JavaFileObject;
-
-import static java.nio.charset.StandardCharsets.UTF_8;
+import javax.tools.SimpleJavaFileObject;
 
 final class TestFiler implements Filer {
+  static class Source extends SimpleJavaFileObject {
+    private final OutputStream stream;
+    protected Source(URI uri, OutputStream stream) {
+      super(uri, Kind.SOURCE);
+      this.stream = stream;
+    }
+    @Override public OutputStream openOutputStream() {
```

Comment:
As an instance of `Source` is only created by `TestFiler#createSourceFile`, I think it's okay to remember the already opened stream here. That instance will be (and according to all the green lights *is*) only used to collect the Java source lines once. I'm 99% sure, that `openOutputStream` is only called a single time per `TestFiler#createSourceFile`.